pyscf_ipu/electron_repulsion/popcint/libcint.py

```diff
@@ -0,0 +1,506 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+# [x] Refactor into {def int2e_sph, def int1e_nuc, ...}. 
+# [ ] Add tile-mapping of integral computation (what is basic unit we tmap? ). 
+# [ ] Consider how to interface this code into nanoDFT. 
+# [ ] Remove hard-coding of tensor (i.e. move shape computation to python/jax.trace). 
+import os
+import pyscf
+import numpy as np 
+import ctypes
+import ctypes
+import numpy
+from pyscf import lib
+from icecream import ic 
+from functools import partial
+import os.path as osp
+import jax 
+import jax.numpy as jnp 
+
+libcgto = numpy.ctypeslib.load_library("libcint.so", "") 
+
+ANG_OF     = 1
+NPRIM_OF   = 2
+NCTR_OF    = 3
+KAPPA_OF   = 4
+PTR_EXP    = 5
+PTR_COEFF  = 6
+BAS_SLOTS  = 8
+NGRIDS     = 11
+PTR_GRIDS  = 12
+
+INT1E_KIN     = 0
+INT1E_NUC     = 1
+INT1E_OVLP    = 2
+INT1E_KIN_IP  = 3
+INT1E_NUC_IP  = 4
+INT1E_OVLP_IP = 5
+INT2E_SPH     = 6
+INT2E_IP1_SPH = 7
+
+def make_loc(bas, key):
+    if 'cart' in key:
+        l = bas[:,ANG_OF]
+        dims = (l+1)*(l+2)//2 * bas[:,NCTR_OF]
+    elif 'sph' in key:
+        dims = (bas[:,ANG_OF]*2+1) * bas[:,NCTR_OF]
+    else:  # spinor
+        l = bas[:,ANG_OF]
+        k = bas[:,KAPPA_OF]
+        dims = (l*4+2) * bas[:,NCTR_OF]
+        dims[k<0] = (l[k<0] * 2 + 2) * bas[k<0,NCTR_OF]
+        dims[k>0] = (l[k>0] * 2    ) * bas[k>0,NCTR_OF]
+
+    ao_loc = numpy.empty(len(dims)+1, dtype=numpy.int32)
+    ao_loc[0] = 0
+    dims.cumsum(dtype=numpy.int32, out=ao_loc[1:])
+    return ao_loc
+
+def getints2c(intor_name, N, atm, bas, env, shls_slice=None, comp=1, hermi=0,
+              ao_loc=None, cintopt=None, out=None):
+    natm = atm.shape[0]
+    nbas = bas.shape[0]
+    shls_slice = (0, nbas, 0, nbas)
+    ao_loc = make_loc(bas, intor_name)
+
+    shape = (N, N, comp)
+    prefix = 'GTO'
+
+    dtype = numpy.double
+    drv_name = prefix + 'int2c'
+
+    #mat = numpy.ndarray(shape, dtype, out, order='F')
+    mat = numpy.zeros(shape, dtype=dtype, order="F")#, dtype, out, order='F')
+    cintopt = None 
+
+    # type 
+    float32 = "#define dtype float" in open("libcint.c", "r").read()
```

Comment:
I assume this is not going to be called a lot?  It will be expensive, and after ~1000 calls will fail due to running out of file handles.

Suggest a `@memoized` function called e.g. `libcint_is_float32()`, which closes the file.

Reply:
Or just compute it up at line 19 where you are already interacting with the filesyste,